nucypher/blockchain/eth/currency.py

```diff
@@ -0,0 +1,197 @@
+        return NU(int(self) * int(other), 'NUWei')
+
+    def __floordiv__(self, other) -> 'NU':
+        return NU(int(self) // int(other), 'NUWei')
+
+    def __gt__(self, other) -> bool:
+        return int(self) > int(other)
+
+    def __ge__(self, other) -> bool:
+        return int(self) >= int(other)
+
+    def __lt__(self, other) -> bool:
+        return int(self) < int(other)
+
+    def __le__(self, other) -> bool:
+        return int(self) <= int(other)
+
+    def __int__(self) -> int:
+        """Cast to smallest denomination"""
+        return int(self.to_nu_wei())
+
+    def __repr__(self) -> str:
+        r = f'{self.__symbol}(value={str(self.__value)})'
+        return r
+
+    def __str__(self) -> str:
+        return f'{str(self.to_tokens())} {self.__symbol}'
+
+
+class Stake:
+    """
+    A quantity of tokens, and staking time-frame for one stake for one miner.
+    """
+
+    __ID_LENGTH = 16
+
+    def __init__(self,
+                 owner_address: str,
+                 index: int,
+                 value: NU,
+                 start_period: int,
+                 end_period: int):
+
+        # Stake Info
+        self.owner_address = owner_address
+        self.index = index
+        self.value = value
+        self.start_period = start_period
+        self.end_period = end_period
+        self.duration = self.end_period - self.start_period
+
+        # Internals
+        self.start_datetime = datetime_at_period(period=start_period)
+        self.end_datetime = datetime_at_period(period=end_period)
+        self.duration_delta = self.end_datetime - self.start_datetime
+
+    def __repr__(self):
+        r = f'Stake(index={self.index}, value={self.value}, end_period={self.end_period})'
+        return r
+
+    def __eq__(self, other):
+        return bool(self.value == other.value)
+
+    @classmethod
+    def from_stake_info(cls, owner_address, index: int, stake_info: Tuple[int, int, int]):
+        """Reads staking values as they exist on the blockchain"""
+        start_period, end_period, value = stake_info
+        instance = cls(owner_address=owner_address,
+                       index=index,
+                       start_period=start_period,
+                       end_period=end_period,
+                       value=NU(value, 'NUWei'))
+        return instance
+
+    def to_stake_info(self) -> Tuple[int, int, int]:
+        """Returns a tuple representing the blockchain record of a stake"""
+        return self.start_period, self.end_period, int(self.value)
+
+    @property
+    def id(self) -> str:
+        """TODO: Unique staking ID, currently unused"""
+        digest = b''
+        digest += eth_utils.to_canonical_address(address=self.owner_address)
+        digest += str(self.index).encode()
+        digest += str(self.start_period).encode()
+        digest += str(self.end_period).encode()
+        digest += str(self.value).encode()
+        stake_id = sha256(digest).hex()[:16]
```

Comment:
This approach for encoding integer values is prone to collisions. Let's focus in the last 4 concatenated items in the digest (i.e., index, start_period, end_period, value). These two stakes (with the same owner_address) have a colliding ID:
* `"1" + "2345" + "6789" + "5555"`
* `"12" + "3456" + "7895" + "555"`
There are many collisions that can be achieved this way. An easy solution is to include a separator character that doesn't occur in any of the variables.

Reply:
Yeah - Thanks for your input on this method.  It is totally unused at the moment, just a drafted Idea.  So - I can just add a delimiter between values to avoid collisions?

Reply:
Yes, as long as that character cannot occur on any of those vars. First three are plain ints, so any char would work, but `str(self.value)` is `NU`, right? Not so sure, here. Perhaps something like `"//"`, I don't know.

Reply:
Perhaps there is no need for this feature?  How about using the pipe character `|` ?

Reply:
:heavy_check_mark: